### ISSUE TITLE
Does not belong to the language

### ISSUE THREAD

**Issue by fregante** [jasnell/proposal-deprecated#5]
Having this as part of the language is not a benefit to the language itself. 

Why should a `deprecated` function be penalized by the engine? It takes 10 minutes for someone to write a search/replace to drop the string in production "and make your code faster"

**Deprecation should only surface during development.** If it works, it should keep working.

There's already a mechanism to mark functions as deprecated during development: JSDoc’s `@deprecated` comments.

**Comment by lveneris** [comment]
The point is not to benefit the language. The point is to benefit everyone using it.

Not to mention your point doesn't really work - nobody has said anything about the engine penalizing `deprecated` functions. What has been said, however, is VMs and tooling should be able to easily identify deprecated functions and display (or throw) a specified warning.

As for your last comment, specifying deprecation in the documentation (which, by the way, not all packages have) is not a catch-all way to notify people. Often people read the code itself instead of reading the documentation to work out what things do; in fact, I myself do that all the time, especially in projects with poor and indescriptive documentation.

Therefore, I think your criticism here is greatly unfair. Having such a feature, particularly as a recognizable standard pragma, would be vastly beneficial and make the language's ecosystem easier to work with.

**Comment by fregante** [comment]
> nobody has said anything about the engine penalizing `deprecated` functions

Yes the readme does: **VM implementations can automatically de-optimize code in any `deprecated` scope**

De-optimizing code means penalizing it, slowing it down.

> which, by the way, not all packages have

I think you misunderstood what [JSDocs](https://jsdoc.app/tags-deprecated.html) is, because it's not "documentation", it's plain comments, which means _any code_ can have it, just like they could have this statement.

Your suggested

```js
function deprecatedFunction() {
  deprecated;
  // do stuff
}
```

is already completely achieved with:

```js
/** @deprecated You can even give a reason */
function deprecatedFunction() {
  // do stuff
}
```

> tooling should be able to easily identify deprecated functions

They already can, without altering the language: 
- https://devblogs.microsoft.com/typescript/announcing-typescript-4-0-beta/#deprecated-support
- https://code.visualstudio.com/updates/v1_49#_deprecated-tag-support-for-javascript-and-typescript

> as a recognizable standard pragma

While JSDoc is not part of the language itself, it has existed for years and more and more tools recognize it and handle it **already.** What does `deprecated` really offer over `@deprecated`? I don't see a single advantage over "but it's not part of the language"

**Comment by parzhitsky** [comment]
BTW, this is how it looks now in VSCode, – pretty helpful already:

<img width="568" alt="image" src="https://user-images.githubusercontent.com/59528510/113509083-4c072a80-955c-11eb-907d-8fc1db59d8e6.png">

**Comment by lveneris** [comment]
I was naive when I wrote that comment, back in November 2020. Oh, how times change.

You're absolutely right. It isn't the scope of the language itself - rather, perhaps Node and other runtimes should issue a warning similar to that of its deprecated unhandled promise exceptions warning when people use an object marked with the `@deprecated` JSDoc directive. The current solutions issued by editors like VSCode are fine, *however*, it would be better if, similar to how NPM issues a warning when you install deprecated packages, a runtime or some other utility in the ecosystem warned you when using deprecated code.

**Comment by fregante** [comment]
I kinda expect browsers to start recognizing the `@deprecated` tag at some point. They already show all kind of deprecation warnings in the console for their own APIs.

**Comment by parzhitsky** [comment]
Visible deprecation information is a subject of stylistic choices of a given environment: VSCode shows it as a crossed-through text with Markdown popup, Node.JS runtime prints a text in stdout with a specific prefix, WebStorm, or IntelliJ, or other envs might do wildly different things.

What we want eventually is an agreed-upon way of  _marking_ something as deprecated, as in “avoid using the ___, it is deprecated now”. In JavaScript, we currently have `@deprecated` pragma is JSDoc (I guess, it is pretty much agreed-upon, correct me if I’m wrong). We can try to generalize it even further, beyond JavaScript, but I’m afraid then it becomes out of the scope of this repo.

**Comment by lveneris** [comment]
> We can try to generalize it even further, beyond JavaScript

That isn't gonna happen, unfortunately. The ways of doing things are different for every language, and thus unless the entire software engineering community can pull together to ask people to start using the agreed-upon best solution, that isn't going to be done.

**Comment by jasnell** [comment]
> perhaps Node and other runtimes should issue a warning similar to that of its deprecated unhandled promise exceptions warning when people use an object marked with the @deprecated JSDoc directive.

The challenge with this is that the information is not surfaced to the runtimes. That is, when v8 compiles the JavaScript, we lose all comment data and cannot see the `@deprecated` indicators at all.

Further, there are times where we deprecate blocks of code *within* a function, such as deprecating a particular pattern of arguments while leaving the rest of the function intact.

Imagine something like...

```js
function foo(a) {
  if (a) {
    /* @deprecated */
  } else {
    // not deprecated
  }
}

/**
 * @deprecated
 */
function bar() {}

foo()  // Not Deprecated

foo(true)  // Deprecated

foo(false) // Not Deprecated

bar()  // Deprecated
```js

Currently it's a bit cumbersome for us to insert the various deprecation warnings at the right spots and we end up with a significant amount of duplicated code and odd idioms.

What I'd like to get to is the ability to simply flag a block of code as deprecated and allow the JS runtime to optionally let us know if it's hit... e.g.

```js
function foo(a) {
  if (a) {
    deprecated;
  } else {
    // not deprecated
  }
}

function boo() {
  deprecated;
}
```

... then have the option of instructing v8 (or whatever other JS engine) to invoke a callback when it hits a `deprecated;` statement in the code, complete with stack trace.

The idea is still very rough, and I can fully accept the position that it may not belong in the language. It would, however, save us a bit of headache over our current mechanism.

**Comment by parzhitsky** [comment]
> deprecating a particular pattern of arguments

Actually, since you can already do this with TypeScript, I think there is (or might eventually be) a way of doing that with JSDoc:

```ts
/** @deprecated Passing user's raw JSON representation is deprecated */
function getUserAge(user: string): number;
function getUserAge(user: User): number;

function getUserAge(arg: string | User): number {
  let user: User;

  if (typeof arg === 'string')
    user = JSON.parse(arg);
  else
    user = arg;

  return user.age;
}
```

<img width="621" alt="image" src="https://user-images.githubusercontent.com/59528510/114175763-03c47f80-9943-11eb-85ad-b5ceb1d1f6d9.png">
